### ISSUE TITLE
Isolating conda environments from ~/.local

### ISSUE THREAD

**Issue by ghost** [conda/conda#7173]
I think, I am seeing a bug in conda create command where I am not able to create an isolated conda environment, which doesn't inherit packages from the base environment, in Ubuntu 16.04.

The reproducible steps for the bug are as follows.

1) Create a new Ubuntu 16.04 VM or any Ubuntu 16.04 machine.

2) Install azure-cli
pip install azure-cli

Let's assume the python version is 2.7 here, as that is the default python in Ubuntu 16.04

This command installs packages ~/.local/lib/python2.7

3) Install miniconda in Ubuntu 16.04

4) conda create -n testenv python=2.7

5) pip freeze

In pip freeze, I see the azure-cli packages from ~/.local/lib/python2.7 , which indicates to me that the testenv conda environment is not isolated, as it inherited the packages from the base environment.

I only checked this for azure-cli, but my guess is that it is true for other packages too.

For python virtual env, this is not an issue, I get an isolated python virtual environment. I also don't get this issue on windows. I only get this on Linux and Mac machines. 

Please take a look at the issue and let me know if I am doing something wrong or it is a genuine bug. 

Thanks

**Comment by kalefranz** [comment]
@mingwandroid What are your latest thoughts here?  CPython has explicit cutouts and code paths to handle virtualenv.  Maybe we should patch our cpython to basically always be in virtualenv mode?  Of course, that's going to make people mad who are accustomed to the current behavior.  Either way, I think there will be a portion of people unhappy.  I'm not sure how to thread this needle.

**Comment by adyada** [comment]
One way I figured out to create an isolated conda environment is to 1) Create a python virtualenv, which is isolated. 2) Then clone that python virtualenv using conda create --clone option. Clone option doesn't have this problem.

**Comment by kalefranz** [comment]
Yeah the virtualenv you created is hooking into that hard-coded cpython behavior.  I think the difference something related to `sys.prefix` vs `sys.exec_prefix`.  Don't remember the exact details without digging into it a bit more.

**Comment by adyada** [comment]
https://github.com/pypa/pip/blob/master/src/pip/_internal/utils/misc.py#L283 is the line where pip has special condition for python virtual envs, which is not true for conda. In case that helps.

**Comment by adyada** [comment]
Yeah, I mean that the testenv conda environment inherited the packages from the system python. I didn't check for the base conda environment.

**Comment by mingwandroid** [comment]
@kalefranz, these deliberate exclusions are part of pip, as @adyada, but really it's [this part](https://github.com/pypa/pip/blob/master/src/pip/_internal/utils/misc.py#L409-L419) that seems to do the avoid `user_site` stuff (which is what `~/.local` is called in the code AFAIK).

We could patch our pip to ignore `user_site`, but half the pip using world would maybe prefer we didn't? Or do you both contend that any exclusion done in `pip` for a virtualenv should also always apply to a conda env? What about the conda base env? Should it still look in `user_site`?

Then there are those who say that conda should never ever look in `user_site`, and I half at least one foot in that camp (maybe we should invent an env var called `CONDA_ECOSYSTEM_ISOLATION` for this purpose? I face the same thing in the R world FWIW).

**Comment by kalefranz** [comment]
> Then there are those who say that conda should never ever look in `user_site`

I personally have a foot there too.  And I'd go so far as saying this should be the default for any conda-installed python (given that any conda-install python will by definition *always* be in a conda environment).  The environment variable I'd propose then would be to open up the interpreter to look in `user_site` if people actually want that behavior.

That's my personal opinion though, and why this issue is marked as a discussion topic.  If we make any type of change here, we need to make sure it's thoroughly discussed and we have consensus on the behavior change.

**Comment by adyada** [comment]
I think, including user_site packages also caused a bug for us, when some people tried installing and using some packages and got errors. That's how the bug was reported to me. I will reproduce that bug and post here by today or tomorrow.

**Comment by adyada** [comment]
The error that I get when user_site packages are also included in an conda environment is that the import of some modules fail, even if that module shows up in pip freeze. 
 I am just pasting the output of pip freeze in the conda env and a command that shows the import failure. 

The important lines are shown as bold lines.

So, backports.tempfile shows up in pip freeze, but then its import fails. 

Apologies for pasting such a long output. Actually, I didn't install most of these packages in the conda environment, these were inherited from the user_site location, that is ~/.local/python2.7

I don't see this issue when I am able to create an isolated python environment using python virtual env

So, I am also on the side that we should not be including user_site packages in a conda env. Probably, might be better to have the conda behavior same as the python virtual env behavior in terms of isolation of environments. 


(testenv) testuser@testVMadyada2: $ python -m pip freeze
adal==0.5.1
applicationinsights==0.11.2
argcomplete==1.9.4
asn1crypto==0.24.0
azure-batch==4.1.1
azure-batch-extensions==1.0.1
azure-cli==2.0.31
azure-cli-acr==2.0.23
azure-cli-acs==2.0.31
azure-cli-advisor==0.5.1
azure-cli-appservice==0.1.31
azure-cli-backup==1.1.1
azure-cli-batch==3.2.0
azure-cli-batchai==0.2.0
azure-cli-billing==0.1.8
azure-cli-cdn==0.0.14
azure-cli-cloud==2.0.13
azure-cli-cognitiveservices==0.1.12
azure-cli-command-modules-nspkg==2.0.1
azure-cli-configure==2.0.15
azure-cli-consumption==0.3.0
azure-cli-container==0.1.22
azure-cli-core==2.0.31
azure-cli-cosmosdb==0.1.20
azure-cli-dla==0.0.19
azure-cli-dls==0.0.21
azure-cli-eventgrid==0.1.12
azure-cli-eventhubs==0.1.2
azure-cli-extension==0.0.12
azure-cli-feedback==2.1.1
azure-cli-find==0.2.9
azure-cli-interactive==0.3.19
azure-cli-iot==0.1.19
azure-cli-keyvault==2.0.21
azure-cli-lab==0.0.21
azure-cli-monitor==0.1.5
azure-cli-network==2.0.28
azure-cli-nspkg==3.0.2
azure-cli-profile==2.0.22
azure-cli-rdbms==0.2.1
azure-cli-redis==0.2.12
azure-cli-reservations==0.1.2
azure-cli-resource==2.0.27
azure-cli-role==2.0.22
azure-cli-servicebus==0.1.2
azure-cli-servicefabric==0.0.12
azure-cli-sql==2.0.25
azure-cli-storage==2.0.31
azure-cli-vm==2.0.30
azure-common==1.1.9
azure-datalake-store==0.0.19
azure-graphrbac==0.40.0
azure-keyvault==0.3.7
azure-mgmt-advisor==1.0.1
azure-mgmt-authorization==0.40.0
azure-mgmt-batch==4.1.0
azure-mgmt-batchai==1.0.0
azure-mgmt-billing==0.2.0
azure-mgmt-cdn==1.0.0
azure-mgmt-cognitiveservices==1.0.0
azure-mgmt-compute==4.0.0rc1
azure-mgmt-consumption==2.0.0
azure-mgmt-containerinstance==0.4.0
azure-mgmt-containerregistry==1.0.1
azure-mgmt-containerservice==3.0.1
azure-mgmt-cosmosdb==0.3.1
azure-mgmt-datalake-analytics==0.2.0
azure-mgmt-datalake-nspkg==2.0.0
azure-mgmt-datalake-store==0.2.0
azure-mgmt-devtestlabs==2.0.0
azure-mgmt-dns==2.0.0rc1
azure-mgmt-eventgrid==0.4.0
azure-mgmt-eventhub==1.2.0
azure-mgmt-iothub==0.4.0
azure-mgmt-iothubprovisioningservices==0.1.0
azure-mgmt-keyvault==0.40.0
azure-mgmt-marketplaceordering==0.1.0
azure-mgmt-monitor==0.5.0
azure-mgmt-msi==0.1.0
azure-mgmt-network==2.0.0rc2
azure-mgmt-nspkg==2.0.0
azure-mgmt-rdbms==1.1.0
azure-mgmt-recoveryservices==0.1.0
azure-mgmt-recoveryservicesbackup==0.1.1
azure-mgmt-redis==4.1.0
azure-mgmt-reservations==0.1.0
azure-mgmt-resource==1.2.1
azure-mgmt-servicebus==0.4.0
azure-mgmt-servicefabric==0.1.0
azure-mgmt-sql==0.8.6
azure-mgmt-storage==1.5.0
azure-mgmt-trafficmanager==0.40.0
azure-mgmt-web==0.35.0
azure-multiapi-storage==0.2.0
azure-nspkg==2.0.0
azure-storage==0.34.3
azureml-core==0.1.0.806005
**backports.tempfile==1.0
backports.weakref==1.0.post1**
bcrypt==3.1.4
certifi==2018.4.16
cffi==1.11.5
chardet==3.0.4
colorama==0.3.9
configparser==3.5.0
cryptography==2.2.2
entrypoints==0.2.3
enum34==1.1.6
funcsigs==1.0.2
futures==3.2.0
humanfriendly==4.10
idna==2.6
ipaddress==1.0.22
isodate==0.6.0
jmespath==0.9.3
jsonpickle==0.9.6
keyring==12.0.1
knack==0.3.2
mock==2.0.0
monotonic==1.4
msrest==0.4.27
msrestazure==0.4.26
ndg-httpsclient==0.4.4
oauthlib==2.0.7
paramiko==2.4.1
pathlib==1.0.1
pathlib2==2.3.0
pathspec==0.5.6
pbr==4.0.2
prompt-toolkit==1.0.15
pyasn1==0.4.2
pycparser==2.18
pydocumentdb==2.3.1
Pygments==2.2.0
PyJWT==1.6.1
PyNaCl==1.2.1
pyOpenSSL==17.5.0
python-dateutil==2.7.2
pytz==2018.4
PyYAML==3.12
requests==2.18.4
requests-oauthlib==0.8.0
ruamel.ordereddict==0.4.13
ruamel.yaml==0.15.37
scandir==1.7
scp==0.10.2
SecretStorage==2.3.1
six==1.11.0
sshtunnel==0.1.3
tabulate==0.8.2
urllib3==1.22
vsts-cd-manager==1.0.1
wcwidth==0.1.7
websocket-client==0.47.0
Whoosh==2.7.4
xmltodict==0.11.0
(testenv) testuser@testVMadyada2: $
(testenv) testuser@testVMadyada2: $
(testenv) testuser@testVMadyada2: $ python
Python 2.7.14 |Anaconda, Inc.| (default, Mar 27 2018, 17:29:31)
[GCC 7.2.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from azureml.core.project import Project
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/testuser/miniconda3/envs/testenv/lib/python2.7/site-packages/azureml/core/project.py", line 15, in <module>
    from azureml.execution.execution_api import ExperimentExecution
  File "/home/testuser/miniconda3/envs/testenv/lib/python2.7/site-packages/azureml/execution/execution_api.py", line 6, in <module>
    from azureml.execution import _commands
  File "/home/testuser/miniconda3/envs/testenv/lib/python2.7/site-packages/azureml/execution/_commands.py", line 37, in <module>
    **from backports import tempfile as temp_dir_back
ImportError: cannot import name tempfile**
>>>

**Comment by kalefranz** [comment]
Is `pip freeze` what's reaching into `~/.local` here then?  I'm guessing `conda list` doesn't show the packages... because those packages aren't actually in `site-packages` within the environment prefix, correct?

**Comment by adyada** [comment]
yeah, conda list doesn't show those packages.

This is the output of conda list.

(testenv) testuser@testVMadyada2:~$ conda list
# packages in environment at /home/testuser/miniconda3/envs/testenv:
#
# Name                    Version                   Build  Channel
azureml-core              0.1.0.806005              <pip>
backports.tempfile        1.0                       <pip>
backports.weakref         1.0.post1                 <pip>
ca-certificates           2018.03.07                    0
certifi                   2018.1.18                py27_0
jsonpickle                0.9.6                     <pip>
libedit                   3.1                  heed3624_0
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 7.2.0                hdf63c60_3
libstdcxx-ng              7.2.0                hdf63c60_3
ncurses                   6.0                  h9df7e31_2
ndg-httpsclient           0.4.4                     <pip>
openssl                   1.0.2o               h20670df_0
pathspec                  0.5.6                     <pip>
pip                       9.0.3                    py27_0
python                    2.7.14              h1571d57_31
readline                  7.0                  ha6073c6_4
ruamel.ordereddict        0.4.13                    <pip>
ruamel.yaml               0.15.37                   <pip>
setuptools                39.0.1                   py27_0
sqlite                    3.22.0               h1bed415_0
tk                        8.6.7                hc745277_3
wheel                     0.31.0                   py27_0
zlib                      1.2.11               ha838bed_2
(testenv) testuser@testVMadyada2:~$

**Comment by adyada** [comment]
So, I could reproduce this behavior on windows too. 

I think, there could be some people who would be depending on user_site packages in a conda environment. So, a solution is to have an argument like "--no-user-site" in the conda create command, which then doesn't include user_site packages. This way conda won't break the behavior for existing users and there would "--no-user-site" argument to not include any user_site packages if someone wants it.

**Comment by mingwandroid** [comment]
>  think, there could be some people who would be depending on user_site packages in a conda environment. So, a solution is to have an argument like "--no-user-site" in the conda create command, which then doesn't include user_site packages. 

The problem is interfacing this conda environment parameter to Python itself which is what is in control of whether `~/.local` is added to the `sys.path`. Using an env. var is sub-optimal for external tooling that does not use the correct `conda activate` procedure. I *guess* we could put this in a file in the root of the environment and then modify our python executable to always look for this? Changing the behaviour of the `python` executable isn't something to do lightly though!

**Comment by kalefranz** [comment]
How does the python executable know if it’s been called from a virtualenv? Maybe we should be flipping that switch for all conda envs too? But then I think that’s going to have other ramifications than just removing .local from sys.path. Definitely something we’d need to be extremely careful about. Actually should probably just reach out to some cpython core maintainers for their thoughts.

**Comment by mbargull** [comment]
xref: https://github.com/pypa/pip/issues/5060

**Comment by mingwandroid** [comment]
I think we would like to be able to configure the root env to also not look in `~/local`, and if so in that case you just want to check a file relative to `sys.prefix` and virtual/base shouldn't matter.

**Comment by jakirkham** [comment]
Please don’t disable `~/.local`. 😟

**Comment by takluyver** [comment]
I'd support disabling user site packages for Python in conda. I always do this in my own conda environments, and I've certainly seen it cause confusion for users, e.g. when a package they just now conda-installed is shadowed by a version they installed with pip ages ago. User site packages is valuable when using system Python installation, to install packages without sudo, but conda is typically often installed in your user account anyway, so it's just an extra complication.

I would do this by setting `ENABLE_USER_SITE = False` in site.py (this is what I set in conda envs locally):

https://github.com/python/cpython/blob/v3.6.5/Lib/site.py#L79-L81

I think that with this set, pip should refuse to do a `--user` install, though I haven't checked whether it currently does so.

**Comment by maphew** [comment]
@jakirkham why not disable `.local` inside a conda env? (real question) What would you put there that wouldn't or couldn't be inside conda `base` and/or `my-project` instead?

**Comment by JustinToribio** [comment]
The workaround I use is I activate a conda environment first, then activate a virtualenv (in that order, it doesn't work if you try activating the virtualenv first and the conda environment second for some reason).  In the Bash terminal, both environments show as active in parentheses, and both "pip freeze" and "conda list" show zero packages (assuming no packages have been installed in either environment).  

I know I could just use the virtualenv, but I prefer using a conda environment.  So this allows me to do that and the sole purpose of the virtualenv is to isolate the conda environment from the global site-packages.

**Comment by adyada** [comment]
Thanks for letting us know the above workaround. This is definitely a more easy workaround. Earlier, I was creating a python virtual env and then cloning that using "conda create --clone" option to get an isolated conda environment, where it was also hard to install multiple python versions.

**Comment by JustinToribio** [comment]
You're welcome.  Through some experimentation I also just stumbled upon some interesting behavior that **allows me to create an isolated conda env on its own, WITHOUT needing to activate an additional `virtualenv`.**  

If I create a new conda env with `conda create --name conda_env`, activate it, don't install any packages, and then run `which jupyter`, I get `~/Anaconda3_May_18_2018/Scripts/jupyter`, and any python scripts or jupyter notebooks I run from `conda_env` are able to import from the global site-packages at `~/Anaconda3_May_18_2018/lib/site-packages`.  

HOWEVER, if I then install into `conda_env` the Jupyter package with `conda install jupyter`, and then run `which jupyter` again, I get `~/Anaconda3_May_18_2018/envs/conda_env/Scripts/jupyter`, and **now any python scripts or jupyter notebooks I run from `conda_env` are no longer able to import from the global site-packages, and they are instead imported from `~/Anaconda3_May_18_2018/envs/conda_env/lib/site-packages`,  which is exactly the behavior I want.**  

Here's the output of "Executing transaction" after running `conda install jupyter` in `conda_env`:  

`DEBUG menuinst_win32:__init__(196): Menu: name: 'Anaconda${PY_VER} ${PLATFORM}', prefix: '~\Anaconda3_May_18_2018\envs\conda_env', env_name: 'conda_env', mode: 'user', used_mode: 'user'
DEBUG menuinst_win32:create(320): Shortcut cmd is ~\Anaconda3_May_18_2018\python.exe, args are ['~\\Anaconda3_May_18_2018\\cwp.py', '~\\Anaconda3_May_18_2018\\envs\\conda_env', '~\\Anaconda3_May_18_2018\\envs\\conda_env\\python.exe', '~\\Anaconda3_May_18_2018\\envs\\conda_env\\Scripts\\jupyter-notebook-script.py', '%USERPROFILE%']`  

It's as if the conda env just needs to be "pointed to" the correct directory of packages.  

This solution works well for me since I will be doing mainly Data Science work (hence, my preference to manage packages and virtual envs with conda), and essentially using Jupyter with each project I work on.  

Hopefully this sheds some light on the issue and a conda envs' ability to isolate from the global site-packages.

**Comment by takluyver** [comment]
I'm revising my views on this. I still disable user site packages in conda environments for myself, but I've come to rely on it for providing packages to users at a research facility.

Specifically, my comment above, that "conda is typically installed in your user account anyway" is not true at DESY/EuXFEL. Anaconda is provided by the facility in a directory that is not writable by users. Users who want to install extra packages can either use `pip install --user`, or create a conda environment in a writable directory.

I'm often working with people who are not that familiar with using Python. They are clever people and could certainly understand environments if they had to, but it's offputting if they have to deal with environments before they can do anything useful, and it's easy to forget to switch to the right environment to run some code. So I tell them to use `pip install --user` to add packages to get stuff up and running. And this relies on conda leaving user-site-packages enabled.

So, for my use cases, I currently want user-writeable environments to be neatly isolated, but non-user-writeable environments to have some kind of per-user overlay, so the user can always install packages into the current environment. I'm not asking anyone to implement that at the moment, but I hope it helps us think about the questions.

**Comment by kilroythethird** [comment]
Is there any progress on this ?
I just run into this and wonder if there is a nice way to have python ignore anything outside the conda environment except the suggested virtualenv inside a conda env or the jupyter notebook workaround.

With the current behavior i am currently NOT able to run psutil in an anaconda env AND have it installed as user.

**Comment by richtong** [comment]
So what is the command line you use to create this virtual env and then clone, is this the new sequence then which would create a venv in the current directory and name it by it's parent directory name. So it works kind of like pipenv.

```
python -m venv .
conda create --clone $(basename ..)
```


> One way I figured out to create an isolated conda environment is to 1) Create a python virtualenv, which is isolated. 2) Then clone that python virtualenv using conda create --clone option. Clone option doesn't have this problem.

**Comment by kousu** [comment]
I have a different workaround than using `python -m virtualenv`, from https://github.com/conda/conda/issues/7173#issuecomment-389190657

```
env=your-new-env-name
conda create $env
sed -i 's/^ENABLE_USER_SITE.*$/ENABLE_USER_SITE = False/' (conda env list | grep "^$env" | awk '{print $NF}')/lib/python*/site.py
```

I feel this form is better for noninteractive installs.

**Comment by javidcf** [comment]
Another possible solution that doesn't involve editing the standard library is setting the [`PYTHONNOUSERSITE`](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONNOUSERSITE) environment variable within the conda environment. [This post](https://stackoverflow.com/questions/31598963) shows two different ways to do that. You can add an activation and a deactivation script to your environment, as explained in the official documentation, [Saving environment variables](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#saving-environment-variables). But it seems from conda 4.8 you can do it even more conveniently through the `conda` command, which now allows you to specify environment variables to set per conda environment, as explained in [Setting environment variables](https://conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#setting-environment-variables).

TLDR: Just run this.

    conda env config vars set PYTHONNOUSERSITE=1 -n my_environment

**Comment by g-kartik** [comment]
Devs please consider this issue. Can't this be solved by modifying how conda does things?

**Comment by mfansler** [comment]
Python implements `-s` and `-E` (and `-I`) flags, providing means for isolating from user space. Is the demand for Conda to provide something else due to limitations of these capabilities or unawareness of these options?

**Comment by kousu** [comment]
I think people are confused that `conda` and `venv` are inconsistent. We're expecting that an "environment" means something with some isolation.

## plain:

```
kousu@ail:/tmp$ /usr/bin/python3 -c 'import sys; print(sys.path)'
['', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/home/kousu/.local/lib/python3.6/site-packages', '/usr/local/lib/python3.6/dist-packages', '/usr/lib/python3/dist-packages']
```


## venv:

```
kousu@ail:/tmp$ python3 -m venv env1
kousu@ail:/tmp$ env1/bin/python3 -c 'import sys; print(sys.path)'
['', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/tmp/env1/lib/python3.6/site-packages']
```

## virutalenv

same results:

```
kousu@ail:/tmp$ virtualenv env2 >/dev/null
kousu@ail:/tmp$ env2/bin/python3 -c 'import sys; print(sys.path)'
['', '/usr/lib/python36.zip', '/usr/lib/python3.6', '/usr/lib/python3.6/lib-dynload', '/tmp/env2/lib/python3.6/site-packages']
```

## conda

```
kousu@ail:/tmp$ ~/miniconda3/bin/conda create -q -p /tmp/env3 -y python=3.6 >/dev/null
kousu@ail:/tmp$ env3/^C
kousu@ail:/tmp$ env3/bin/python3 -c 'import sys; print(sys.path)'
['', '/tmp/env3/lib/python36.zip', '/tmp/env3/lib/python3.6', '/tmp/env3/lib/python3.6/lib-dynload', '/home/kousu/.local/lib/python3.6/site-packages', '/tmp/env3/lib/python3.6/site-packages']
```


conda is _replacing_ the system `/usr/lib` paths with its own isolated versions, but _including_ `~/.local/lib/python3.6/site-packages` by default, so its environment is only half isolated. It's just kinda confusing. https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html says 

>      With conda, you can create, export, list, remove, and update environments that have different versions of Python and/or packages installed in them.


You're right, `-s` makes it behave like `venv`:

```
kousu@ail:/tmp$ env3/bin/python3 -s -c 'import sys; print(sys.path)'
['', '/tmp/env3/lib/python36.zip', '/tmp/env3/lib/python3.6', '/tmp/env3/lib/python3.6/lib-dynload', '/tmp/env3/lib/python3.6/site-packages']
```

but it's difficult to make every script/`entry_point` in an env use that. Should we do something messy like

```
find env3/bin -name "*.py" | xargs sed 's/python/python -s/'
```

? Why can't we just have the environment work?

So, I think the request is something like enabling `-s` by default in conda envs to match `venv`.

**Comment by kousu** [comment]
Poking around a little bit, reading `env3/lib/python3.6/site.py`, I learned how `venv` pulls it off: it's natively supported by cpython:

> If a file named "pyvenv.cfg" exists one directory above sys.executable,
> sys.prefix and sys.exec_prefix are set to that directory and
> it is also checked for site-packages (sys.base_prefix and
> sys.base_exec_prefix will always be the "real" prefixes of the Python
> installation). If "pyvenv.cfg" (a bootstrap configuration file) contains
> the key "include-system-site-packages" set to anything other than "false"
> (case-insensitive), the system-level prefixes will still also be
> searched for site-packages; otherwise they won't.


and taking a guess tried

```
kousu@ail:/tmp$ echo "include-system-site-packages = false" > env3/pyvenv.cfg 
kousu@ail:/tmp$ env3/bin/python3 -c 'import sys; print(sys.path)'
['', '/tmp/env3/lib/python36.zip', '/tmp/env3/lib/python3.6', '/tmp/env3/lib/python3.6/lib-dynload', '/tmp/env3/lib/python3.6/site-packages']
```

So the request is: please make `conda create -p some/path python` create `some/path/pyvenv.cfg` when it installs.

I can see why this is a bit weird: conda outgrew its python roots so there's no good place to put this feature. You'll need to go fiddle with the install scripts for your python packages, *not* conda itself. But lots of us still use it as a virtualenv alternative.

**Comment by less-and-less-bugs** [comment]
I met this question just a few days ago. I solved it as follows after I had created an unisolated environment:
1.  `python -m site `
     to get sys.path information  of this environment. I found site.ENABLE_USER_SITE was Ture as default that meat packages 
     in ./local would be seen by Module search algorithm of python. 

2. `python -m site -h` 
      to get path of site.py 
3.   `vim path of site.py`
       Then set ENABLE_USER_SITE False. Then all the packages in ./local would not be found by used environment now

**Comment by RafalSkolasinski** [comment]
Is there any current plan to deal with this? Every now and then this issue hits one by surprise. 

I believe that if one creates an environment, conda or virtualenv - whichever, they expect a full isolation from other python installations.  If that is not provided this lead to hard to reproduce errors:
- one may think that the environment definition is sufficient and complete
- behaviour may be different on other machines
- unpredictable behaviour

Most of my use-cases is to create environments using conda (I basically use it as easy environment manager to create / activate them) and then I do all python install inside using `pip`. 

I really thought that I finally solved issue with conda environments being polluted by site-packages by adding
```
create_default_packages:
  - conda-ecosystem-user-package-isolation
```
to my `~/.condarc` but  just now for some reason on one machine it did not work as expected and lead to long debugging...

**Comment by RafalSkolasinski** [comment]
> @RafalSkolasinski didn't know that package existed! star_struck

:)  Yes, I don't even remember how I found it - I think it may have been recommended to me on gitter / reddit when I complained about the issue looong time ago. 

> I notice it doesn't do anything to insulate against use of [PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH),

Interesting, it may be what was hitting me now. Cannot say for sure - it was quite a fresh install so I went on rampage of removal / reinstall of things until it stopped happening...

**Comment by less-and-less-bugs** [comment]
> 

In fact, I met this question when I used tmux at a Linux server. I indeed tried some methods on GitHub but it did not work. At last, I found use "conda deactivate " first to leave the base python environment provided by conda, and then went into tmux session. It was successful!

**Comment by mfansler** [comment]
> > I notice it doesn't do anything to insulate against use of [PYTHONPATH](https://docs.python.org/3/using/cmdline.html#envvar-PYTHONPATH),
> 
> Interesting, it may be what was hitting me now. Cannot say for sure - it was quite a fresh install so I went on rampage of removal / reinstall of things until it stopped happening...

@RafalSkolasinski FWIW, not knowing about the aforementioned package, I had previously created [my own personal version of it](https://github.com/mfansler/envvar-pythonnousersite-true) that only sets the `PYTHONNOUSERSITE=1`. Since it's basically a template for setting an environment variable, I added [a similar one for masking `PYTHONPATH`](https://github.com/mfansler/envvar-pythonpath-null).

Both are available via Anaconda Cloud, and have the MatchSpec:

```
merv::envvar-pythonnousersite-true
merv::envvar-pythonpath-null
```

The latter could be a possible alternative if the `conda-ecosystem-user-package-isolation` doesn't want to include PYTHONPATH masking.

**Comment by kousu** [comment]
@less-and-less-bugs wrote:

>     3. `vim path of site.py`
>        Then set ENABLE_USER_SITE False. Then all the packages in ./local would not be found by used environment now

This is exactly what we used to do:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/e6eb4299fa7d9e4a2fb8f6362c26908847a49ec8/install_sct#L608

But then I figured out (https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/3606) that this was a little more reliable:

https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/7198415f186385c9d3d2e4246d58da4b0bf7e262/install_sct#L605

So maybe that'll help you.

I still am surprised every time that conda *doesn't* do this by default, since it's main feature (at least, as seen by the zeitgeist) is to provide reproducibility.

**Comment by niander** [comment]
This creates issues when using `tox-conda` plugin since it will use conda to create an environment but pip to install dependencies. Unfortunately, I am not sure how to inject a `pyvenv.cfg` when tox is handling the environments.

**Comment by mfansler** [comment]
@niander what is the problem exactly? Using `pip` is fine - it defaults to installing into the active Python's prefix. The issue is `pip install --user` installs. Is that what `tox` is using?

**Comment by niander** [comment]
No, they use regular pip install. From logs I can see that first `tox-conda` create a conda environment. Then, it installs `pytest` because that is one of the dependencies in my testenv.  `pytest` have `attrs >= 19.3.0` as dependency but pip doesn't install it because there is already a newer version of `attrs` installed in the user site-packages location. Since this is Windows, the location is `%APPDATA%\roaming\python\python37\site-packages`.

However, when later `tox` calls `pytest`, it will fail with ModuleNotFoundException.
So `pip` found the package in site-packages but when calling `pytest` it can't find it anymore.

**Comment by mfansler** [comment]
Okay, makes sense. Have you tried including `conda-ecosystem-user-package-isolation` package under `conda_deps`? That should isolate the Python installation from user space.

**Comment by niander** [comment]
@mfansler I tried but then lates I got an issue with calling pip install directly inside a toxenv (I do this to install a local package in dev mode). pip complains it can't find ssl library.

Nonetheless, I am still failing to understand why when pytest executes it can't find `attr` if this is installed in the site-packages. If I start the python inside the conda environment, I can do `import attr`, but the pytest executed by `tox` doesn't. I am confused by that.

**Comment by niander** [comment]
I figured it out. It is because of the way `tox` executes commands. The subprocess that tox creates to execute the test commands doesn't include the environment `%APPDATA%`.

The code in `site.py` looks for the user site-packages and if that exists adds that to the python path (unless it is a venv, but not applicable here since it is a conda env and there is no pyvenv.cfg). The logic in `site.py` for Windows is to see if `%APPDATA%` exists, if not, uses `~` (`%USERPROFILE%`). 

So, in this case, running inside tox, because the real user site-packages exists only inside `%APPDATA%` and not `~`, `site.py` will assume that there is no user site-packages and won't add it to the python path.

What I am not sure is why when `tox` (or `tox-conda`) is just installing dependencies, the env variables are different so the user site-packages is found and added to python path.

**Comment by itcarroll** [comment]
So this issue has been ongoing for a long time, and github bot keeps closing issues that are not resolved:
- #394
- #448
- #7707
- maybe others?

Workarounds described include the [conda-ecosystem-user-package-isolation](https://anaconda.org/conda-forge/conda-ecosystem-user-package-isolation) package or setting `PYTHONNOUSERSITE` manually. Here is a case where this workaround is not useful.

A command line tool installed outside a conda environment, using `pip install --user`, goes in `~/.local/bin/` and includes as its first line `#!/path/to/python`. *THAT* Python is outside the conda environment. With `PYTHONNOUSERSITE` however, site packages are disabled for *ALL* Python interpreters. This breaks the command line tool.

New workaround.

Although conda does not create a *virtual* environment, a `pyvenv.cfg` file is checked by the `site` module regardless. So, do what venv does. Create a `pyvenv.cfg` file for your conda environment with the line `include-system-site-packages = false`.

```
$ conda activate <name>
$ cat > $CONDA_PREFIX/pyvenv.cfg
include-system-site-packages = false
$
```

I don't know what other repercussions that has involving `sys.prefix`, but so far so good. Your mileage may vary.

I strongly feel the conda dev team should provide the `--no-site-packages` option requested a decade ago in #394.

**Comment by mfansler** [comment]
> Although conda does not create a virtual environment, a `pyvenv.cfg` file is checked by the site module regardless. So, do what `venv` does. Create a `pyvenv.cfg` file for your conda environment with the line `include-system-site-packages = false`.

This sounds like an excellent solution. The code for that mechanism has been stable since Python v3.3.

This could also be implemented as a Conda Forge package that literally just delivers a `pyvenv.cfg` to the `$PREFIX` with that line. I'd be happy to support it (feel free to tag me on [staged-recipes](https://github.com/conda-forge/staged-recipes) to review).

**Comment by jezdez** [comment]
@itcarroll Please open a new ticket with your summary, it's no use piling on this 5-year-old ticket with yet more discussion where most of the current maintainers were not around.

**Comment by jezdez** [comment]
Full disclosure, I've opened https://github.com/conda/infrastructure/pull/850 to lock older issue sooner, to make sure we're able to encourage actionable issues instead of reviving old ones.

**Comment by itcarroll** [comment]
@jezdez Done! You never know whether maintainers are more averse to duplication or re-activation, so thank you for the clear guidance.
